Pod setup restart fixes and improvements (Loop issue #2117)

itsmojo opened the pull request.

itsmojo committed:
Add FirstAppear

itsmojo committed:
Alphabetize OmniKitUI/Views files

itsmojo committed:
Add cannulaInsertionSuccessfullyStarted var to podState & pumpManager

itsmojo committed:
For an in-progress insert cannula compute a wait time on remaining bolus

itsmojo committed:
Call new resetPerPodPumpManagerState() func when starting a pod
Reset podAttachmentConfirmed state variable for each new pod
Reset all per pod PumpManagerAlerts for each new pod
Update PumpStatusHighlight to "Finish Setup" while activating a paired pod
Add comment noting that podCommState.deactivating value is never returned
Add some missing state initializion for simulator in jumpStartPod()

itsmojo committed:
UI fixes and improvements for various pod setup restart situations
+ Rename func & enum names from pair to pairAndPrime for clarity
+ Start with pairAndPrime state when prime has not finalized
+ Gracefully handle restarts when prime or insert cannula is in progress
+ Add restart support to PairPodViewModel & View to skip button wait
+ Add restart support to InsertCannulaViewModel & View to skip button wait
+ Fix checkCannulaInsertionFinished() to use checkingInsertion state
+ Fix PodPairer.podIsActivated to properly handle a paired pod
+ Add new PodPairer.getPodCommState var to implement podIsActivated
+ Add missing LocalizedString for "Insert Cannula" navigation bar title

marionbarker commented:
Successful test with Eros pod.

Please see OmniBLE test notes as well at: https://github.com/LoopKit/OmniBLE/pull/113#issuecomment-1935182660

Tested with Eros pod using an SE phone running iOS 17.3.
Configuration:

* LoopWorkspace-dev, commit 7340d55
* OmniBLE, dev branch commit 8f78c34
* OmniKit, main branch commit f2fc341
* Modified Pod.swift cannulaInsertionUnitsExtra from 0.0 to 5.0 to make the insertion process last longer.

Test during Pair/Priming:
* Touch Pair and wait for pod to begin priming
* quit Loop app, restart Loop app, touch the Finish Setup icon on main screen
   * pod was still priming, returns to Pair Pod screen, reconnects and shows Priming progress bar (repeated 2 times)
* quit Loop app, restart Loop app, touch the Finish Setup icon on main screen
   * pod was finished priming, returns to Pair Pod screen, reconnects and shows the check Paired with Continue button available

Test autolock disable
* wait 30 sec between taps when moving through screens to verify phone does not lock

Test during insertion
* Code was modified to insert 5.5 units before completing insertion
   * So insertion will last about 110 sec
* start insertion, wait for clicking and progress bar
* quit Loop app, restart Loop app, touch the Finish Setup icon on main screen
   * pod was still giving insertion bolus (clicking), returns to Insert Cannula screen, reconnects and shows Inserting progress bar (repeated 2 times)
* quit Loop app, restart Loop app, touch the Finish Setup icon on main screen
   * pod was finished with insertion bolus, returns to Insert Cannula screen, reconnects and shows the check Inserted with Continue button available

itsmojo committed:
Updated and improved resetPerPodPumpManagerState func logic
+ Use retractAlert func to handle the state.activeAlerts removal
+ Reset the state.acknowledgedTimeOffsetAlert var in case it gets used
+ Add call to resetPerPodPumpManagerState func from forgetPod()
Simplified restart logic for insert cannula view restarts

marionbarker commented:
Code review:
* Compared code for last commit of this PR with code for last commit of OmniBLE 113
* Code changes are consistent

See testing performed with DASH pod at https://github.com/LoopKit/OmniBLE/pull/113#issuecomment-1947555014

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -435,7 +435,8 @@ public class PodCommsSession {
             if status.podProgressStatus == .insertingCannula {
                 podState.setupProgress = .cannulaInserting
                 podState.updateFromStatusResponse(status, at: currentDate)
-                return insertionWait // Not sure when it started, wait full time to be sure
```
Is `insertionWait` no longer used? Maybe it should be removed above.

itsmojo replied:
Removed insertionWait and replaced its one remaining use with a calculation based on status return

ps2 commented on OmniKitUI/ViewControllers/OmnipodUICoordinator.swift:
```diff
@@ -214,6 +214,11 @@ class OmnipodUICoordinator: UINavigationController, PumpManagerOnboarding, Compl
                 self?.navigateTo(.deactivate)
             }
 
+            if pumpManager.podCommState == .activating {
```
This logic seems like view model logic. Can you expose the needed api `pumpManager.podCommState` from the `PodPairer` protocol, and move this logic into the viewmodel?

itsmojo replied:
Done using your suggested updates

ps2 commented on OmniKitUI/ViewModels/InsertCannulaViewModel.swift:
```diff
@@ -134,6 +135,13 @@ class InsertCannulaViewModel: ObservableObject, Identifiable {
         }
     }
     
+    var restarting: Bool {
```
I found this name "restarting" to be confusing, as I couldn't figure out what was being restarted. Is "resuming" maybe a more descriptive name? They can't actually restart pairing.  They're just trying to resume where they left off, right?

itsmojo replied:
This is all gone now using your suggested updates

ps2 commented on OmniKitUI/Views/InsertCannulaView.swift:
```diff
@@ -76,9 +76,13 @@ struct InsertCannulaView: View {
         .navigationBarItems(trailing: cancelButton)
+        .onFirstAppear {
+            // handle possible restart without waiting for a button action
+            self.viewModel.handlePossibleRestart()
```
This also should be view model logic, and not a view triggered action. That will obviate the need for the custom `onFirstAppear` ViewModifier.

itsmojo replied:
Done using your suggested updates

ps2 reviewed (changes requested):
This is an important fix, thanks.  It just seems like some of the view model logic is misplaced in the navigator and views.

itsmojo committed:
Keep logic around resuming progress indicators in the view model
+ Remove cannulaInsertionSuccessfullyStarted var from pumpManager
+ Remove use of insertionWait variable from PodCommsSession.insertCannula()
+ Remove .restarting enum from PairPadViewModel & InsertCannulaViewModels
+ Remove restarting testing and viewModel state setting in UICoordinator
+ Remove restarting Bool var from PairPadViewModel & InsertCannulaViewModels
+ Remove handlePossibleRestart() from PairPadViewModel & InsertCannulViewModel
+ Remove use of onFirstAppear and delete now unneeded FirstAppear.swift
+ Improved PodPairer to use the pumpManager podCOmmState var
+ Update CannulaInserter to implement cannulaInsertionSuccessfullyStarted var
+ Update PairPodViewModelState priming value to use an optional
+ Have the view model init() funcs test for and handle resume cases
